Add test cases for NETFramework netstandard2 support

ericstj opened the pull request:
This adds test cases for validating inclusion of netstandard2 support libraries when targeting desktop.

Fixes #1287 

Scenarios not yet covered: non-SDK projects (PackageReference, project.json, packages.config)

In doing this work I discovered one issue which impacts the ValueTuple scenario: my exclusions were missing a trailing ')' causing us to select the wrong ValueTuple assembly (both were added and the first one was selected which was wrong for net47).

Note that for the net47/ValueTuple test to build correctly the machine requires the net47 targeting pack to be installed.  I'm not sure if this will be the case on all machines so we might want to guard that with a check for the targeting pack.

/cc @dsplaisted @nguerrera

ericstj committed:
Ensure Net47 specific support libraries are included

Previously the Exclude tag had a typo causing net461 files to be preferred.

ericstj committed:
Add test cases for NS2.0 netframework support

ericstj committed:
Add header to Extensions targets and include in MSBuildAllProjects

dsplaisted commented:
To fix the `All_props_and_targets_add_themselves_to_MSBuildAllTargets` test, make sure all of the new props / targets files that have been added have the following:

```xml
  <PropertyGroup>
    <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
  </PropertyGroup>
```

ericstj commented:
Yeah, I have a local fix for that as well as some other tests.  Hold off on review for a moment.

ericstj committed:
Make restore work for Packages.config tests

Packages.config projects need to be restored with nuget.exe.

ericstj committed:
Make ValueTuple test only run in 4.7 targeting pack is installed

ericstj commented:
This should be good now.  PTAL.

ericstj commented:
Looking into the full framework failures.

ericstj commented:
To fix the sequencing issue I'm going to further simplify the sequencing logic for the extensions _HandlePackageFileConflicts.  Previously I was careful to not set AfterTargets="ResolveNuGetPackageAssets" because that could cause log spew about an unknown target to appear in projects that didn't have the nuget targets.  That's no longer a concern since we're only installing to VS versions that have the target.  This lets me collapse the last two cases to be the same.

ericstj committed:
Fix sequencing of _HandlePackageFileConflicts

Our previous sequencing assumed a later import (as was the case when
this came from a nuget package) and also assumed that nuget's targets
weren't necessarily imported.

To work with an earlier import we cannot make use of properties like
ResolveNuGetPackages and ProjectLockFile, but since we can depend on
NuGet targets being present we can schedule after ResolveNuGetPackageAssets
without causing log spew about missing targets.

As a result we can schedule ourselves Before RAR and After ResolveNuGetPackageAssets.
This ensures we'll run before RAR and as close to ResolveNuGetPackageAssets as
possible (if it runs).

ericstj commented:
Ok, this is ready to go.  @dsplaisted @nguerrera can one of you review please?

dsplaisted commented:
@MattGertz for approval.  This PR includes tests and bugfixes to the .NET Framework support for .NET Standard.  Original template is [here](https://github.com/dotnet/sdk/pull/1279#issuecomment-305942903)

dsplaisted commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopExeWtihNetStandardLib.cs:
```diff
@@ -0,0 +1,334 @@
+        public void It_includes_netstandard(bool isSdk, ReferenceScenario scenario)
+        {
+            var testAsset = _testAssetsManager
+                .CopyTestAsset(GetTemplateName(isSdk), identifier: scenario.ToString())
```
I think the identifier should include bot the value of `isSdk` as well as the `scenario` in order to use a unique path for each case of the theory.

ericstj replied:
They use different templates which get included in the path.  I was actually hitting max-path at one point which is what I shortened the test method names and tried to be brief in these identifiers.

dsplaisted commented on test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildADesktopLibrary.cs:
```diff
@@ -91,6 +92,113 @@ public void Method1()
+        [WindowsOnlyFact]
+        public void It_can_use_ValueTuple_and_exchange_the_type_with_a_NETStandard_library()
+        {
+            var referenceAssemblies = ToolLocationHelper.GetPathToDotNetFrameworkReferenceAssemblies(TargetDotNetFrameworkVersion.Version47);
```
I don't think we'll get any automated coverage of this.  Are the .NET 4.7 reference assemblies part of the VS install?  They aren't on my machine.

ericstj replied:
I downloaded and installed it.  It's an optional component starting with 15.1.  Do you think we can install it on the CI machines?  Should be as simple as running https://www.microsoft.com/en-us/download/details.aspx?id=55168.
